### PR TITLE
Add documentation for memory primitives

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: doc/architectural/memory-primitives.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+pointers as having 64 bits. This can be changed by various options (see section
+"C/C++ frontend options" in the output of `cbmc --help`).
+
+Memory is represented in cbmc as a set of objects. Each object represents a
+contiguous sequence of bytes and is identified via a numeric object ID. For
+example, assuming integers of width 4 and chars of with 1, a global integer
+variable would correspond to an object of size 4, and memory allocated via
+`malloc(10)` would correspond to an object of size 10.
+
+A pointer then consists of two parts: the upper n bits form the object ID, and
```
Comment (NlightNFotis) [inline doc/architectural/memory-primitives.md:24]: could this be changed to `n`?

**Review by NlightNFotis — APPROVED** [review]
I think it's good to go.

My only preference would be to pick either `cbmc` or `CBMC` and stick with it for all references to it for consistency's sake.

File: doc/architectural/memory-primitives.md (hunk 1 above)
Comment (danpoe) [reply doc/architectural/memory-primitives.md:24]: Here the n doesn't refer to code or commandline text or something like that so I think it shouldn't be highlighted that way.

File: doc/architectural/memory-primitives.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
+\ingroup module_hidden 
+\page memory-primitives Memory Primitives
+
+This document describes the semantics and usage of memory-related and
+pointer-related primitives in cbmc.
```
Comment (peterschrammel) [inline doc/architectural/memory-primitives.md]:
```suggestion
pointer-related primitives in CBMC.
```

File: doc/architectural/memory-primitives.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+
+# Background
+
+
+## Memory and pointers in cbmc
```
Comment (peterschrammel) [inline doc/architectural/memory-primitives.md]:
```suggestion
## Memory and Pointers in CBMC
```

File: doc/architectural/memory-primitives.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+When cbmc analyzes a program, by default it uses the architectural parameters of
+the platform it is running on. That is, on a 64-bit system, cbmc will treat
```
Comment (peterschrammel) [inline doc/architectural/memory-primitives.md]:
```suggestion
the platform it is running on. That is, on a 64-bit system, CBMC will treat
```
etc

File: doc/architectural/memory-primitives.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+__CPROVER_assume(i < 0);
+int result = rectify(i);
+assert(result == 0);
+```
+
+Here, the value of `i` is nondeterministically chosen from all the possible
+integer values, and then constrained to negative values via the assumption.
+In cbmc, like uninitialized integers, uninitialized pointers are treated as
+having a nondeterministic value. That is, the value of the pointer itself is
+nondeterministically chosen, though **no memory is allocated**.
```
Comment (peterschrammel) [inline doc/architectural/memory-primitives.md]: Maybe say that you are going to describe what can be expected from an intitialized pointer below.

File: doc/architectural/memory-primitives.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+if it points to a live memory object. That is, it points to the start or to
+somewhere within the sequence of bytes that makes up the memory object.
+
+Conversely, a pointer is invalid if it is null, uninitialized,  points to
+deallocated dynamic memory, points to an out-of-scope local variable, or has a
+value that does not point to (dynamically, automatically, or statically)
+allocated memory, or is out of bounds of the memory object it points to (i.e.,
+the memory object identified by `__CPROVER_POINTER_OBJECT(p)`).
+
+
```
Comment (peterschrammel) [inline doc/architectural/memory-primitives.md:140]: Maybe say upfront that some operations, particularly on invalid pointers, are unspecified. Say how I can detect that I have something with unspecified semantics in my harness.

**Review by peterschrammel — APPROVED** [review]

File: doc/architectural/memory-primitives.md (hunk 4 above)
Comment (danpoe) [reply doc/architectural/memory-primitives.md]: Done

File: doc/architectural/memory-primitives.md (hunk 6 above)
Comment (danpoe) [reply doc/architectural/memory-primitives.md:140]: Done

File: doc/architectural/memory-primitives.md (hunk 5 above)
Comment (danpoe) [reply doc/architectural/memory-primitives.md]: Added a sentence saying that pointers should be explicitely initialized

**Pull request merged**